### PR TITLE
[AutoParallel] Save&Load Module

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

Others
### Describe

* Add save and load module to support autoparallel. 
* Let the parallelizer to use a new distributed context. 
* Set the final distributed context to the default context in the paralleizer.

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/utils.py [hunk 1]
```diff
@@ -314,3 +316,95 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
                 var, tensor_dist_attr)
             var._set_attr('dim_mapping' + core.kAutoParallelSuffix(),
                           dim_mapping)
+
+
+def check_append_sd(append_sd):
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/utils.py]: check_append_sd -> check_append_state_dict ?
append_sd -> append_state_dict ?

File: python/paddle/distributed/auto_parallel/utils.py [hunk 2]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+                                      "model_state_rank{}.pdmodel".format(rank))
+
+        state_dict = {
+            "model": program.state_dict(),
+            "ranks": paddle.distributed.get_world_size()
+        }
+        if _check_addition_info(addition_info):
+            state_dict["addition_info"] = addition_info
+        paddle.save(state_dict, ckpt_file_name)
+        print("Successfully saving model to {}".format(output_dir))
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/utils.py]: Would it be better to use logging or paddle log instead of print?

File: python/paddle/distributed/auto_parallel/utils.py [hunk 3]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+                        format(str(type(addition_info))))
+    else:
+        return addition_info
+
+
+def _check_valid_dir(file_dir):
+    """
+    Validity check of input directory
+    """
+    if file_dir is None:
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/utils.py]: check the file_dir is "" instead of None? because if the file_dir is "", u will return [""]

File: python/paddle/distributed/auto_parallel/utils.py [hunk 4]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
                 var, tensor_dist_attr)
             var._set_attr('dim_mapping' + core.kAutoParallelSuffix(),
                           dim_mapping)
+
+
+def _check_addition_info(addition_info):
+    """
+    Validity check of additional information
+    """
+    if addition_info is None:
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/utils.py]: please check addition_info is {}

File: python/paddle/distributed/auto_parallel/utils.py [hunk 5]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+        return addition_info
+    elif not isinstance(addition_info, dict):
+        raise TypeError("The type of addition_info should be dict, but got {}".
+                        format(str(type(addition_info))))
+    else:
+        return addition_info
+
+
+def _check_valid_dir(file_dir):
```
Comment (Caozhou1995) [inline python/paddle/distributed/auto_parallel/utils.py]: dir means the directory, Would it be better to name it path？

File: python/paddle/distributed/auto_parallel/utils.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

File: python/paddle/distributed/auto_parallel/utils.py [hunk 6]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+        output_dir(str): The path of the checkpoint file to be saved.
+        is_integrated(bool, optional): Whether to integrate param before save. Default: False.
+        addition_info(dict, optional): Additional information. Default: None.
+        dist_attr_dir(str, optional): The path of distributed attribute file to be saved. Default: None
+
+    Returns:
+        None
+
+    Example:
+        >>> output_dir = os.path.join(args.output_dir, "step_%d" % step)
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/utils.py]: I think `>>>` marks are no need. Please follow the criterion.

File: python/paddle/distributed/auto_parallel/utils.py [hunk 7]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+            raise ValueError("The type of each directory should be str.")
+        if not all(os.path.exists(file) for file in file_dir):
+            raise ValueError("The directory's file does not exist.")
+        return file_dir
+    else:
+        raise TypeError("The type of directory should be list, but got {}.".
+                        format(str(type(file_dir))))
+
+
+def save_static_checkpoint(program,
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/utils.py]: Why do we name this function as `static`?

File: python/paddle/distributed/auto_parallel/utils.py [hunk 8]
```diff
@@ -357,3 +362,127 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+        >>> save_static_checkpoint(program, output_dir)
+    """
+    if not is_integrated:
+        rank = paddle.distributed.get_rank()
+        ckpt_file_name = os.path.join(output_dir,
+                                      "model_state_rank{}.pdmodel".format(rank))
+
+        state_dict = {
+            "model": program.state_dict(),
+            "ranks": paddle.distributed.get_world_size()
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/utils.py:436]: The `get_world_size` returns only one number of used ranks, why do we use `ranks`?

**Review by aoyulong — COMMENTED** [review]
It is ok that the save&load reuse the original save&load of Paddle. But the distributed information should be saved properly. You next pr should enhance the function and test it.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 2 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 3 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 4 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 8 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py:436]: The "ranks" attribution is for checking whether the user input the complete param when merging and slicing parameters.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 7 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: Done. The function name has been changed to "save_distributed_checkpoint".

File: python/paddle/distributed/auto_parallel/utils.py (hunk 6 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 5 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py]: done.

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_auto_parallel_save_load PROPERTIES TIMEOUT 120)`

File: python/paddle/distributed/auto_parallel/utils.py [hunk 9]
```diff
@@ -352,3 +360,140 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+        program(Program): The program to be saved.
+        checkpoint_path(str): The path of the checkpoint file to be saved.
+        is_integrated(bool, optional): Whether to integrate param before save. Default: False.
+        addition_info(dict, optional): Additional information. Default: None.
+        dist_attr_path(str, optional): The path of distributed attribute file to be saved. Default: None
+
+    Returns:
+        None
+
+    Examples:
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/utils.py:422]: example should contain usage for addition_info

File: python/paddle/distributed/auto_parallel/utils.py [hunk 10]
```diff
@@ -352,3 +360,140 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+
+        state_dict = {
+            "model": program.state_dict(),
+            "ranks": paddle.distributed.get_world_size()
+        }
+        if _check_addition_info(addition_info):
+            state_dict["addition_info"] = addition_info
+
+        paddle.save(state_dict, ckpt_file_name)
+        logging.info("Already save model to {}".format(checkpoint_path))
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/utils.py:442]: save --> saved

File: python/paddle/distributed/auto_parallel/utils.py [hunk 11]
```diff
@@ -352,3 +360,140 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+                                dist_attr_path=None):
+    """ 
+    Save model parameter state, optimzer state, distributed attribute and 
+    additional information of each rank.
+
+    Args:
+        program(Program): The program to be saved.
+        checkpoint_path(str): The path of the checkpoint file to be saved.
+        is_integrated(bool, optional): Whether to integrate param before save. Default: False.
+        addition_info(dict, optional): Additional information. Default: None.
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/utils.py:416]: the document for addition_info is too simple and make user puzzled. should given a hint about what is the "addition_info" meaning in document, like the training step / learning rate, etc

File: python/paddle/distributed/auto_parallel/utils.py [hunk 12]
```diff
@@ -352,3 +360,140 @@ def make_data_unshard(dist_main_prog, dist_startup_prog):
...
+    checkpoint_path = _check_valid_path(checkpoint_path)
+    dist_attr_path = _check_valid_path(dist_attr_path)
+
+    if checkpoint_path and dist_attr_path:
+        raise NotImplementedError(
+            "Merge&Slice parameter with dist_attr has not been implemented.")
+
+    elif checkpoint_path:
+        assert len(checkpoint_path) == paddle.distributed.get_world_size(), \
+            "The number of checkpoint_path must equal to the number of ranks"
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/utils.py:486]: the prompt should be that: "if dist_attr is not provided, the number of checkpoint_path should equal to the number of current world size.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 8 above)
Comment (JZ-LIANG) [reply python/paddle/distributed/auto_parallel/utils.py:436]: it should be named as "world size"

**Review by JZ-LIANG — APPROVED** [review]

File: python/paddle/distributed/auto_parallel/utils.py (hunk 12 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py:486]: ok. It will be updated in the next pr.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 11 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py:416]: ok. It will be updated in the next pr.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 10 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py:442]: ok. It will be updated in the next pr.

File: python/paddle/distributed/auto_parallel/utils.py (hunk 9 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/utils.py:422]: ok. It will be updated in the next pr.
